examples/camera-app/linux/src/camera-device.cpp

```diff
@@ -37,6 +39,176 @@ using namespace chip::app::Clusters::CameraAvSettingsUserLevelManagement;
+{
+    CSI,
+    USB,
+    Other,
+    Unkown,
+};
+
+CameraType detectCameraType(const std::string & fullDevicePath)
+{
+    if (!std::filesystem::exists(fullDevicePath))
+    {
+        return CameraType::Unkown;
+    }
+
+    std::string videoDeviceName = std::filesystem::path(fullDevicePath).filename(); // ex: video0
+    std::string sysPath         = "/sys/class/video4linux/" + videoDeviceName + "/device/driver/module";
+
+    char resolvedPath[PATH_MAX];
+    ssize_t len = readlink(sysPath.c_str(), resolvedPath, sizeof(resolvedPath) - 1);
+
+    // Define driver name constants to avoid magic strings
+    constexpr const char * kCsiDriver1 = "bm2835";
+    constexpr const char * kCsiDriver2 = "unicam";
+    constexpr const char * kUsbDriver  = "uvc";
+
+    if (len != -1)
+    {
+        resolvedPath[len] = '\0';
+        std::string driverPath(resolvedPath);
+
+        if (driverPath.find(kCsiDriver1) != std::string::npos || driverPath.find(kCsiDriver2) != std::string::npos)
+        {
+            return CameraType::CSI;
+        }
+        else if (driverPath.find(kUsbDriver) != std::string::npos)
+        {
+            return CameraType::USB;
+        }
+        else
+        {
+            return CameraType::Other;
+        }
+    }
+
+    return CameraType::Unkown;
+}
+
+GstElement * CreateSnapshotPipelineV4l2(const std::string & device, int width, int height, int quality, int framerate,
+                                        const std::string & filename, CameraError & error)
+{
+    GstElement *pipeline, *source, *jpeg_caps, *videorate, *videorate_caps, *queue, *filesink;
+
+    // Create the pipeline elements
+    pipeline = gst_pipeline_new("snapshot-pipeline");
+    // TODO: Have the video source passed in.
+    source         = gst_element_factory_make("v4l2src", "source");
+    jpeg_caps      = gst_element_factory_make("capsfilter", "jpeg_caps");
+    videorate      = gst_element_factory_make("videorate", "videorate");
+    videorate_caps = gst_element_factory_make("capsfilter", "timelapse_framerate");
+    queue          = gst_element_factory_make("queue", "queue");
+    filesink       = gst_element_factory_make("multifilesink", "sink");
+
+    if (!pipeline || !source || !jpeg_caps || !videorate || !videorate_caps || !queue || !filesink)
+    {
+        ChipLogError(Camera, "Not all elements could be created.");
+
+        // Unreference the elements that were created
+        unrefGstElements(pipeline, source, jpeg_caps, videorate, videorate_caps, queue, filesink);
+
+        error = CameraError::ERROR_INIT_FAILED;
+        return nullptr;
+    }
+
+    // Set the source device and caps
+    g_object_set(source, "device", device.c_str(), "do-timestamp", TRUE, nullptr);
+
+    GstCaps * caps = gst_caps_new_simple("image/jpeg", "width", G_TYPE_INT, width, "height", G_TYPE_INT, height, "framerate",
+                                         GST_TYPE_FRACTION, framerate, 1, "quality", G_TYPE_INT, quality, nullptr);
```

Comment:
I wonder if using `//` could be used to make the formatter make these more readable.

```suggestion
    GstCaps * caps = gst_caps_new_simple(
        "image/jpeg",                                 //
        "width", G_TYPE_INT, width,                   //
        "height", G_TYPE_INT, height,                 //
        "framerate", GST_TYPE_FRACTION, framerate, 1, //
        "quality", G_TYPE_INT, quality,               //
        nullptr
    );
```